packetbeat/protos/mysql/mysql.go

```diff
@@ -667,8 +785,161 @@ func (mysql *mysqlPlugin) receivedMysqlResponse(msg *mysqlMessage) {
 	mysql.publishTransaction(trans)
 	mysql.transactions.Delete(trans.tuple.Hashable())
 
-	logp.Debug("mysql", "Mysql transaction completed: %s", trans.mysql)
-	logp.Debug("mysql", "%s", trans.responseRaw)
+	logp.Debug("mysql", "Mysql transaction completed: %s %s %s", trans.query, trans.param, trans.mysql)
+	//	logp.Debug("mysql", "%s", trans.responseRaw)
+}
+
+func (mysql *mysqlPlugin) parseMysqlExecuteStatement(data []byte, stmtdata *mysqlStmtData) []string {
+
+	var paramType, paramUnsigned uint8
+	nparamType := []uint8{}
+	paramString := []string{}
+	nparam := stmtdata.numOfParameters
+	offset := 0
+	// mysql hdr
+	offset += 4
+	// cmd type
+	offset++
+	// stmt id
+	offset += 4
+	// flags
+	offset++
+	// iterations
+	offset += 4
+	// null-bitmap
+	if nparam > 0 {
+		offset += (nparam + 7) / 8
+	}
+	// stmt bound
+	stmtBound := data[offset]
+	offset++
+	paramOffset := offset
+	if stmtBound == 1 {
+		paramOffset += nparam * 2
```

Comment:
Probably you can add a check that `len(data) >= paramOffset` here.

Reply:
protocol fuzzing test or malicious attack will cause the exception

Reply:
+1 on offset check.

Packetbeat being a passive sniffer also has to deal with packet loss and being started late. That is, protocol analyzers in Packetbeat need  to sync and re-sync every now and so often to the actual TCP stream. Not being in sync leads to parsing errors. The other protocols drop state on parsing error and retry. Until we eventually find a packet starting with a query. Depending on query rates (lower rates are more likely to sync on), this can take a while. In order to not generate panics for the time the analyzer is not in sync yet, better have the analyzer somewhat robust.

Reply:
+1

Reply:
Actually when calling this function, it has guaranteed this is a full length mysql packet according to the mysql header.But offset check will better have the analyzer robust.